src/providers/ad/ad_gpo_child_utils.c

```diff
@@ -0,0 +1,252 @@
+#include "util/debug.h"
+#include "util/atomic_io.h"
+
+#define INI_GENERAL_SECTION "General"
+#define GPT_INI_VERSION "Version"
+#define GPT_INI "/GPT.INI"
+
+static errno_t parse_ini_file_with_libini(struct ini_cfgobj *ini_config,
+                                          int *_gpt_version)
+{
+    int ret;
+    struct value_obj *vobj = NULL;
+    int gpt_version;
+
+    ret = ini_get_config_valueobj(INI_GENERAL_SECTION, GPT_INI_VERSION,
+                                  ini_config, INI_GET_FIRST_VALUE, &vobj);
+    if (ret != 0) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "ini_get_config_valueobj failed [%d][%s]\n", ret, strerror(ret));
+        goto done;
+    }
+    if (vobj == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "section/name not found: [%s][%s]\n",
+              INI_GENERAL_SECTION, GPT_INI_VERSION);
+        ret = EINVAL;
+        goto done;
+    }
+
+    gpt_version = ini_get_int32_config_value(vobj, 0, -1, &ret);
+    if (ret != 0) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "ini_get_int32_config_value failed [%d][%s]\n",
+              ret, strerror(ret));
+        goto done;
+    }
+
+    *_gpt_version = gpt_version;
+
+    ret = EOK;
+
+ done:
+
+    return ret;
+}
+
+static errno_t gpo_sanitize_buffer_content(uint8_t *buf, int buflen)
+{
+    int i;
+    int line_start = 0;
+    int equal_pos = 0;
+
+    if (!buf) {
+        return EINVAL;
+    }
+
+    for (i = 0; i < buflen; ++i) {
+        if (buf[i] == '\n') {
+            line_start = i + 1;
+            continue;
+        }
+        if (buf[i] == '=') {
+            equal_pos = i;
+            continue;
+        }
+        if (isascii(buf[i])) {
+            continue;
+        }
+
+        /* non-ascii */
+        if (equal_pos <= line_start) { /* key */
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Key or section starting at position %d ('%.*s...') contains"
+                  " non-ascii symbol. File is unusable!\n",
+                  line_start, i - line_start, buf + line_start);
+            return EINVAL;
+        }
+
+        buf[i] = '?';
+        DEBUG(SSSDBG_IMPORTANT_INFO,
+              "Value for key '%.*s' contains non-ascii symbol."
+              " Replacing with '?'\n",
+              equal_pos - line_start, buf + line_start);
+    }
+
+    return EOK;
+}
+
+/*
+ * This function parses the GPT_INI file stored in the gpo_cache, and uses the
+ * results to populate the output parameters ...
+ */
+errno_t ad_gpo_parse_ini_file(const char *smb_path, int *_gpt_version)
+{
+    struct ini_cfgfile *file_ctx = NULL;
+    struct ini_cfgobj *ini_config = NULL;
+    const char *ini_filename;
+    int ret;
+    int gpt_version = -1;
+    TALLOC_CTX *tmp_ctx = NULL;
+    struct stat st;
+    int fd = -1;
+    uint8_t *buf = NULL;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ini_filename = talloc_asprintf(tmp_ctx, GPO_CACHE_PATH"%s%s",
```

Comment:
I'd like to propose you another approach to solve this problem, but let me step back a bit first.

From my point of view the line that I have underlined doesn't follow any good practices convention. I'd expect this line to concatenate two strings: the path and the file. But instead it concatenates three lines. Shouldn't we move the `GPO_CACHE_PATH` and `smb_path` concatenation one level above? I mean to the caller of the `ad_gpo_parse_ini_file()` function. This would lower the level of coupling and it would ease the development of tests and any other function calling it.

This in turn would help you to get the test running as the caller for the production environment can concatenate the prefix `GPO_CACHE_PATH` while the test environment can use any other string (or no string at all).

This is just an idea and it might be wrong, but in my experience many problems happening during the unit testing phase come from a bad design.

Reply:
> I'd expect this line to concatenate two strings: the path and the file.

What do you call a "file"?

But in general, right, getting prefix out of this function is a viable solution, that I considered.

I just have to postpone this now. Feel free to take over if you'd like.

Reply:
> What do you call a "file"?

I meant the file name, in this case `GPT_INI`.

> I just have to postpone this now. Feel free to take over if you'd like.

I'll try to have a look at it next week.

Reply:
I think I fixed it. [Code location](https://github.com/ikerexxe/sssd/commits/gpo-encoding-test). How do you want to proceed with it?

Reply:
I cherry-picked your patch. Thanks.

Reply:
I think that if you remove the `reverse_path()` function the tests will pass.

Reply:
Now it fails during `ci-make-distcheck`.

Reply:
The error is the [following](https://github.com/SSSD/sssd/runs/5660348579?check_suite_focus=true):
```
../src/tests/cmocka/test_ad_gpo.c:409:17: error: unused variable 'path' [-Werror=unused-variable]
  409 |     const char *path;
```

Do I remove it in my commit? Or do you do that? I'd like to avoid having many commits in this PR.

Reply:
Updated, but I doubt this is a reason of `ci-make-distcheck` fail:
https://s3.eu-central-1.amazonaws.com/sssd-ci/PR-6065/3/debian10/ci-make-distcheck.log
  --  it was a 'warning' there.

Reply:
> Updated, but I doubt this is a reason of `ci-make-distcheck` fail

Unfortunately and as expected, it was not.

Reply:
`GPT.INI` file was missing from the [build target](https://github.com/ikerexxe/sssd/commit/8c681e52ddba05a84d66824cec6895df89223eed). I've updated the [branch](https://github.com/ikerexxe/sssd/commits/gpo-encoding-test) with that commit.

Reply:
> I'm lost as to why the distcheck is failing right now. I'm running it on my computer and the test passes.

Probably because I didn't update PR yet.

Reply:
My bad, I thought you had updated it.

Reply:
Updated.